hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -117,10 +117,22 @@ object If {
     if (cnsq.typ == altr.typ)
       If(cond, cnsq, altr)
     else {
-      val t = unifyType.getOrElse(cnsq.typ.deepOptional())
-      If(cond,
-        PruneDeadFields.upcast(cnsq, t),
-        PruneDeadFields.upcast(altr, t))
+      cnsq match {
+        case NA(_) => If(cond, NA(altr.typ), altr)
+        case Die(msg, _) => If(cond, Die(msg, altr.typ), altr)
+        case Literal(_, value) => If(cond, Literal(altr.typ, value), altr)
+        case _ =>
+          altr match {
+            case NA(_) => If(cond, cnsq, NA(cnsq.typ))
+            case Die(msg, _) => If(cond, cnsq, Die(msg, cnsq.typ))
+            case Literal(_, value) => If(cond, cnsq, Literal(cnsq.typ, value))
+            case _ =>
+              val t = unifyType.getOrElse(cnsq.typ.deepOptional())
+              If(cond,
+                PruneDeadFields.upcast(cnsq, t),
+                PruneDeadFields.upcast(altr, t))
```

Comment:
Instead of adding the special cases above, could we fix this just by modifying `upcast`? For example, make `upcast(NA(_), t) = NA(t)`. That feels more correct.

Reply:
I don't think this works (we talked about this right)